File: src/ignore/stale-detector.ts
import type Database from 'better-sqlite3';
import { getFunctionBySignatureHash } from '../store/index-store.js';
import type { IgnoreFile, StaleWarning } from './types.js';

export function detectStaleExclusions(
  db: Database.Database,
  ignoreFile: IgnoreFile,
): StaleWarning[] {
  const warnings: StaleWarning[] = [];

  for (let i = 0; i < ignoreFile.exclusions.length; i++) {
    const exclusion = ignoreFile.exclusions[i];

    for (const side of ['a', 'b'] as const) {
      const ref = exclusion.pair[side];
      const stored = getFunctionBySignatureHash(db, ref.signatureHash);

      if (!stored) {
        warnings.push({
          exclusionIndex: i,
          side,
          functionName: ref.function,
          path: ref.path,
          reason: `Function "${ref.function}" in ${ref.path} no longer exists or its signature has changed`,
        });
      }
    }
  }

  return warnings;
}

File: src/ignore/stale_detector.rs
use crate::ignore::types::{IgnoreFile, StaleWarning};
use rusqlite::Connection;

use crate::store::index_store::get_function_by_signature_hash;

pub fn detect_stale_exclusions(
    conn: &Connection,
    ignore_file: &IgnoreFile,
) -> Vec<StaleWarning> {
    let mut warnings = Vec::new();

    for (i, exclusion) in ignore_file.exclusions.iter().enumerate() {
        for (side_char, side) in [('a', &exclusion.pair.a), ('b', &exclusion.pair.b)] {
            if get_function_by_signature_hash(conn, &side.signature_hash).is_none() {
                warnings.push(StaleWarning {
                    exclusion_index: i,
                    side: side_char,
                    function_name: side.function.clone(),
                    path: side.path.clone(),
                    reason: format!(
                        "Function \"{}\" in {} no longer exists or its signature has changed",
                        side.function, side.path
                    ),
                });
            }
        }
    }

    warnings
}
